### PR TITLE
Add pagination support to publicRooms

### PR DESCRIPTION
**Pull request by erikjohnston**
Changes:
 - Remove support for secondary directory servers, now that clients can request remote directories themselves.
 - `start` and `end` keys have been removed from the response to `/publicRooms`, and replaced by `next_batch` and `prev_batch` (whose existence depends on if there is a next or prev batch).
 - The rooms are now returned in order of number of users.
 - `/publicRooms` accept `limit` and `since` parameters.
 - Clients can paginate forwards and backwards by passing the `{next,prev}_batch` tokens to `since=`. There is no `direction` parameter.
 - The room list is not updated with newly rooms published rooms when specifying a `since` token. `new_rooms` is returned in the response when a since token is specified and is a boolean indicating if there are new rooms. Unpublished rooms do get removed, however.
 - The `limit` param is advisory, in particular synapse will return fewer rooms if they have been unpublished.
 - Tokens are not valid indefinitely. Synapse return an error if the tokens are over a month old.

The response now looks something like:

```json
{
    "chunk": [
        {
            "canonical_alias": "#wiiiiiiiiimble:localhost:8480",
            "world_readable": false,
            "num_joined_members": 1,
            "room_id": "!AyxdOWyQKnFhBrCyfS:localhost:8480",
            "guest_can_join": false,
            "aliases": [
                "#wiiiiiiiiimble:localhost:8480"
            ]
        }
    ],
    "prev_batch": "hKFwA6FzzQYfoWTCoW4I",
    "new_rooms": false
}
```

**Question**: If we don't specify a limit, should we return the full list (current behaviour) or should we cap it?

### CODE REVIEW

**Comment by NegativeMjark** [comment]
LGTM

File: synapse/storage/room.py [hunk 1]
```diff
@@ -207,3 +255,71 @@ def add_event_report(self, room_id, event_id, user_id, reason, content,
...
+
+    def get_published_at_stream_id_txn(self, txn, stream_id):
+        sql = ("""
+            SELECT room_id, visibility FROM public_room_list_stream
+            INNER JOIN (
+                SELECT room_id, max(stream_id) AS stream_id
+                FROM public_room_list_stream
+                WHERE stream_id <= ?
+                GROUP BY room_id
+            ) grouped USING (room_id, stream_id)
```
Comment (NegativeMjark) [inline synapse/storage/room.py]: Will this work on postgres, or will you need to use window functions there?

**Comment by erikjohnston** [comment]
@matrixbot retest this please

File: synapse/storage/room.py (hunk 1 above)
Comment (NegativeMjark) [inline synapse/storage/room.py]: Not that it really matters since the table will probably be small to start with.

Comment (erikjohnston) [inline synapse/storage/room.py]: I've been testing on postgres, and it appears to use the right indices.

**Comment by NegativeMjark** [comment]
LGTM

**Pull request merged**